src/iterator.cr

```diff
@@ -381,84 +379,74 @@ module Iterator(T)
   #
-  # def flatten
-  #   Flatten(self, typeof(Flatten.element_type(self))).new(self)
-  # end
-
-  # private class Flatten(I, T1)
-  #   include Iterator(T1)
-
-  #   @iterator : I
-  #   @top : Bool
-  #   @to_rewind : Array(Proc(Nil))
-
-  #   # @generator : I
-
-  #   def initialize(@iterator)
-  #     @generator = @iterator
-  #     @top = true
-  #     @to_rewind = [] of Proc(Nil)
-  #   end
-
-  #   def next
-  #     value = @generator.next
-  #     if value.is_a?(Stop)
-  #       if @top
-  #         return stop
-  #       else
-  #         @generator = @iterator
-  #         @top = true
-  #         return self.next
-  #       end
-  #     end
-
-  #     flatten value
-  #   end
-
-  #   def make_rewinder(iter)
-  #     ->{
-  #       iter.rewind
-  #       # Return nil to disguise the individual iterator types
-  #       nil
-  #     }
-  #   end
-
-  #   def rewind
-  #     @iterator.rewind
-  #     @generator = @iterator
-  #     @top = true
-  #     @to_rewind.each &.call
-  #     @to_rewind.clear
-  #   end
-
-  #   def flatten(element)
-  #     case element
-  #     when Iterator
-  #       flat = element.flatten
-  #       @generator = flat
-  #       @to_rewind << make_rewinder flat
-  #       @top = false
-  #       self.next
-  #     when Iterable
-  #       flatten element.each
-  #     else
-  #       element
-  #     end
-  #   end
-
-  #   def self.element_type(element)
-  #     case element
-  #     when Stop
-  #       raise ""
-  #     when Iterator
-  #       element_type(element.next)
-  #     when Iterable
-  #       element_type(element.each)
-  #     else
-  #       element
-  #     end
-  #   end
-  # end
+  def flatten
+    Flatten(typeof(Flatten.iterator_type(self)), typeof(Flatten.element_type(self))).new(self)
+  end
+
+  private class Flatten(I, T)
+    include Iterator(T)
+
+    @iterator : I
+    @stopped : Array(I)
+    @generators : Array(I)
+
+    def initialize(@iterator)
+      @generators = [@iterator]
+      @stopped = [] of I
+    end
+
+    def next
+      case value = @generators.last.next
+      when Iterator
+        @generators.push value
+        self.next
+      when Iterable
+        @generators.push value.each
+        self.next
+      when Stop
+        if @generators.size == 1
+          stop
+        else
+          @stopped << @generators.pop
+          self.next
+        end
+      else
+        value
+      end
+    end
+
+    def rewind
+      @generators.each &.rewind
+      @generators.clear
+      @generators << @iterator
+      @stopped.each &.rewind
+      @stopped.clear
+    end
+
+    def self.element_type(element)
+      case element
+      when Stop
+        raise ""
+      when Iterator
+        element_type(element.next)
+      when Iterable
+        element_type(element.each)
+      else
+        element
+      end
+    end
+
+    def self.iterator_type(iter)
+      case iter
+      when Iterator
+        [iter, iterator_type(iter.next)].sample
```

Comment:
An easy way to combine two types (get a union of two expressions) is to use `||`, I think (didn't try it) that this can work too:

```cr
iter || iterator_type(iter.next)
```